sink/datadog.go

```diff
@@ -141,13 +159,69 @@ func NewDatadog(monitorId string, opts, tags map[string]string, httpClient *http
 
 		c := datadog.NewConfiguration()
 		c.HTTPClient = httpClient
+		c.Compress = d.compress
 		metricsApi := datadogV2.NewMetricsApi(datadog.NewAPIClient(c))
 		d.metricsApi = metricsApi
 	}
 
 	return d, nil
 }
 
+func (s *Datadog) sendApi(ddCtx context.Context, dp []datadogV2.MetricSeries) error {
+	// Send metrics to Datadog. The Datadog client handles everything; we just pass
+	// it data points.
+	payload := datadogV2.MetricPayload{
+		Series: dp,
+	}
+	_, r, err := s.metricsApi.SubmitMetrics(ddCtx, payload, *datadogV2.NewSubmitMetricsOptionalParameters())
+	// Check to see if the request was too large
+	if r.StatusCode == http.StatusRequestEntityTooLarge {
+		oldPartitionSize := s.partitionSize
+
+		// Sample and try and partition this to fit
+		if pErr := s.updatePartitionSize(dp[0], len(dp)); pErr != nil {
+			return fmt.Errorf("Unable to compute paritions properly send metrics: %s: %s", pErr, err)
+		}
+
+		localPartitionSize := s.partitionSize
+
+		if (localPartitionSize == len(dp) && oldPartitionSize == localPartitionSize) || oldPartitionSize == localPartitionSize {
+			// The partition size already matches the size of the data we have, so we can't retry any more
+			return fmt.Errorf("Unable to send metrics due to size: %s", err)
+		}
+
+		// Repartition and send again
+		var pErr error
+		rangeStart := 0
+		var i int
+		for i = localPartitionSize; i < len(dp); i += localPartitionSize {
+			if err = s.sendApi(ddCtx, dp[rangeStart:i]); err != nil {
+				pErr = fmt.Errorf("%s\n%s", pErr, err)
+			}
+			rangeStart = i
+		}
+
+		if i-rangeStart > 0 {
+			if i > len(dp) {
+				i = len(dp)
+			}
+
+			if err = s.sendApi(ddCtx, dp[rangeStart:i]); err != nil {
+				pErr = fmt.Errorf("%s\n%s", pErr, err)
+			}
+		}
+
+		return pErr
+	}
+
+	if err != nil {
```

Comment:
This needs to be handled before checking `r.StatusCode` because `err` will be a low-level error (e.g. network issue), so there's no `*http.Response`. Afaik, if err is not nil, then there must be an `*http.Response`.